envs robotics - touch sensors - Shadow Hand

ndrwmlnk opened the pull request:
@matthiasplappert
We (@ndrwmlnk @llach) covered the hand model with 92 touch sensors and extended the HandManipulate{Block, Egg, Pen}-v0 environments to HandManipulate{Block, Egg, Pen}TouchSensors-v0. Now, an observation additionally contains 92 touch values (0 - no touch, 1 - touch detected). Sensory information allows faster and better convergence (preliminary results).

Test the environments with pre-trained weights (currently in training):
https://rebrand.ly/TouchSensors
python baselines/her/experiment/play.py HandManipulateBlockTouchSensors.pkl
python baselines/her/experiment/play.py HandManipulateEggTouchSensors.pkl
python baselines/her/experiment/play.py HandManipulatePenTouchSensors.pkl

Touch sensor sites are tailored to physical geoms (robot0:DC_Hand), thus may visually overlap with meshes (robot0:D_Vizual). Active touch sensors are highlighted in red, inactive sensors - in green:
https://rebrand.ly/TouchSensors
sensors_cube_geoms.mp4
sensors_cube_mesh.mp4
sensors_cube.mp4
sensors.mp4

matthiasplappert commented:
Thanks for your PR! I'm excited to get this merged and will work with @pzhokhov to figure out the best next steps.

matthiasplappert commented on gym/envs/robotics/assets/hand/shared.xml:
```diff
@@ -19,9 +19,13 @@
         </default>
         <default class="robot0:DC_Hand">
             <geom material="robot0:MatColl" contype="1" conaffinity="0" group="4"></geom>
+            <!--visualize physical bodies, hide ghost mesh-->
```
Why was this change necessary?

matthiasplappert commented on gym/envs/robotics/hand/manipulate_touch_sensors.py:
```diff
@@ -0,0 +1,146 @@
+        self.touch_visualisation = touch_visualisation
+        self.touch_get_obs = touch_get_obs
+        self._tsensor_prefix = 'robot0:TS_'
+        # self._tsensor_group_names = []  # list of dictionaries of sensors per joint
```
Please remove these commented-out lines.

matthiasplappert commented on gym/envs/__init__.py:
```diff
@@ -377,6 +377,13 @@ def _merge(a, b):
         max_episode_steps=100,
     )
 
+    register(
```
Could you please also register the environments for `Egg` and `Pen` with touch sensing here?

matthiasplappert commented on gym/envs/robotics/hand/manipulate_touch_sensors.py:
```diff
@@ -0,0 +1,146 @@
+# Ensure we get the path separator correct on windows
+MANIPULATE_BLOCK_XML = os.path.join('hand', 'manipulate_block_touch_sensors_76.xml')
+MANIPULATE_EGG_XML = os.path.join('hand', 'manipulate_egg.xml')
+MANIPULATE_PEN_XML = os.path.join('hand', 'manipulate_pen.xml')
```
Would be nice to have touch sensor support here as well. If not, please remove them, since it is otherwise confusing.

matthiasplappert commented on gym/envs/robotics/hand/manipulate_touch_sensors.py:
```diff
@@ -0,0 +1,146 @@
+    raise error.DependencyNotInstalled("{}. (HINT: you need to install mujoco_py, and also perform the setup instructions here: https://github.com/openai/mujoco-py/.)".format(e))
+
+
+def quat_from_angle_and_axis(angle, axis):
```
Please re-use the implementation from `manipulate.py` for this function

matthiasplappert commented on gym/envs/robotics/hand/manipulate_touch_sensors.py:
```diff
@@ -0,0 +1,146 @@
+
+        for k, v in self.sim.model._sensor_id2name.items():  # get touch sensor ids and their site names
+            if 'TS' in v:
+                # self._tsensor_name2id[v] = k
```
Please remove this line.

matthiasplappert commented on gym/envs/robotics/hand/manipulate_touch_sensors.py:
```diff
@@ -0,0 +1,146 @@
+            'desired_goal': self.goal.ravel().copy(),
+        }
+
+    def set_object_size(self, factor):  # changes size of the manipulated object
```
Please move this implementation to the parent class since it's equally applicable to the non-touch version.

matthiasplappert commented on gym/envs/robotics/hand/manipulate_touch_sensors.py:
```diff
@@ -0,0 +1,146 @@
+                touch_values = np.log(np.array(touch_values) + 1.0)
+        observation = np.concatenate([robot_qpos, robot_qvel, object_qvel, touch_values, achieved_goal])
+
+        # set rgba values
```
Please move everything related to visualization to `_render_callback` (https://github.com/openai/gym/blob/master/gym/envs/robotics/robot_env.py#L152) instead. Don't forget to call `super` since the hand environment itself does some visualization here as well.

matthiasplappert commented on gym/envs/robotics/hand/manipulate_touch_sensors.py:
```diff
@@ -0,0 +1,146 @@
+        elif self.touch_visualisation == 'on_touch':
+            for k, v in self._tsensor_id2name.items():
+                if self.sim.data.sensordata[k] != 0.0:
+                    # self.sim.model.site_rgba[self._tsensor_id2siteid[k]] = self._site_id2intial_rgba[self._tsensor_id2siteid[k]].copy()
```
Please remove this line.

pzhokhov commented:
my 2 cents here are the following - a couple of tests of the new environments are failing (tried running manually, unfortunately, those won't show in travis because this is an external PR and as such our mujoco key is not injected => no tests of mujoco-based envs are run). Please run them manually (via `pytest .`  with MUJOCO_KEY env variable set up and make sure 1) tests pass and 2) test count is 359 (that's how much it counted for me when I tried cloning the PR branch and running tests there). If there are some failures that seem unrelated to your changes, let's work together and figure out what's wrong

ndrwmlnk commented:
@matthiasplappert @pzhokhov
Thanks for reviewing the code and tips with pytest .

I improved the code and fixed requested changes. 376 tests passed without errors. The number of tests exceeds 359, probably because of the new touch-sensory robotic environments.
The reason for the previous pytest error was the length of the observation vector:
https://github.com/openai/gym/blob/49cd48020f6760630a7317cb3529a22de6f12f2e/gym/envs/tests/test_envs.py#L23

I added obs to ManipulateTouchSensorsEnv.__init__ here:
https://github.com/ndrwmlnk/gym/blob/0d71b27103cd434173eb24ca50a99b3968866cfd/gym/envs/robotics/hand/manipulate_touch_sensors.py#L59
which now overrides obs from RobotEnv.__init__ here (without touch values):
https://github.com/openai/gym/blob/49cd48020f6760630a7317cb3529a22de6f12f2e/gym/envs/robotics/robot_env.py#L38

Please let me know if I can make any further improvements.

pzhokhov commented:
@matthiasplappert @ndrwmlnk confirmed - tests are passing; if @matthiasplappert is happy, I am happy :)

pzhokhov commented:
here's a successful travis build with the changes: https://travis-ci.org/openai/gym/jobs/489766042 (pay the "All checks have failed" no heed, I messed something up with github webhook when trying to set a build with the changes up)

matthiasplappert commented:
Will take another pass tomorrow (crazy busy week so far) and then we can hopefully get this merged. Thanks everyone for helping!

matthiasplappert commented:
And thanks for working with us on this PR; happy that we got it merged :)
